### DISCUSSION TITLE
有个数据库连接设计的问题想请教大家

### DISCUSSION THREAD

**Discussion by JavenLaw** [cloudwu/skynet#1963] [Q&A] [answered]
我有一个db服务，并且想要与数据库建立n条连接

方案一：
	只开启一个db服务，在db服务里面建立多条连接，例如：
		db(独立服务) ：
						client_1 = mongo.client()
						client_2 = mongo.client()
						client_3 = mongo.client()

方案二：
	开启多个db服务，每个服务只有1条连接，例如：
		db_1(独立服务) ： client = mongo.client()
		db_2(独立服务) ： client = mongo.client()
		db_3(独立服务) ： client = mongo.client()

有点疑问的是：

	例如：db服务有很多消息，都需要去数据库查询/写入。 当db服务处理消息A时，调用db.testcoll:safe_insert()接口

	假设db服务调用db.testcoll:safe_insert()时间很长，那么db服务是一直等待数据返回，不处理后续的B,C消息，还是会继续先处理后续的B,C消息呢？

	即：调用数据库接口，会让当前服务中的任务挂起吗？（会跟服务处理消息的机制一样吗？）

不知道哪个方案更好呢？谢谢大家！

**Comment by huahua132** [comment, marked as answer]
开多个服务吧，多开可以利用到多核优势。
调用db只是协程挂起了，后续请求还会处理

**Reply by JavenLaw** [reply]
嗯嗯，确实多开可以利用多核优势
不过也可能单个服务需要多条连接的情形
例如：一个db服务需要连接2个不同的mongoDB实例（分别在不同的服务器上）
现在知道调用db只是协程挂起，就可以放心建立连接了

**Reply by sniper00** [reply]
> 开多个服务吧，多开可以利用到多核优势。 调用db只是协程挂起了，后续请求还会处理

多个服务需要考虑 同一个key的顺序性 和 多个关联操作的原子性

**Reply by huahua132** [reply]
嗯嗯，是要考虑这些，我觉得可以在调用方处理，比如按key映射到对应服务
